jax/lax/lax.py

```diff
@@ -1378,8 +1378,18 @@ def full(shape: Shape, fill_value: Array, dtype: Optional[DType] = None) -> Arra
     fill_value = xla.device_put_p.bind(convert_element_type(fill_value, dtype))
   return broadcast(fill_value, shape)
 
+# TODO(mattjj) remove, we only need this function because _device_put_raw is
+#  used directly in jnp.array/full instead of device_put_p.bind
+def _is_numpy_eval_tracing():
+  if not config.omnistaging_enabled: return False
+  stack = core.thread_local_state.trace_state.trace_stack
+  # TODO(mattjj) improve:
+  trace_type = stack.stack[0].trace_type
+  # __name__ avoids cyclic dependency with numpy_eval:
+  return trace_type.__name__ == 'NumpyEvalTrace'
+
 def _device_put_raw(x):
-  if isinstance(x, xla.DeviceArray):
+  if isinstance(x, xla.DeviceArray) or _is_numpy_eval_tracing():
```

Comment:
Should we make this mechanism extensible? We could add a registry for non-DeviceArray handling, dependent on the type of the eval trace. On a higher level, it'd be really cool in future if we could have a simple guide explaining how to add a new eval backend.

Reply:
@j-towns I think a better way to allow this in general would be by overriding `device_put_p` and consistently using it in `lax`. The current hack is only required because `np.array` and `np.full` use `_device_put_raw` directly.